File: frameworks/node/koa/Implementation/routes/cors.ts
// rb:wiring cors.*
import cors from "@koa/cors";

import type { Routes } from "../app.ts";
import { fresh } from "../serial.ts";

/**
 * cors: @koa/cors as route middleware on /cors/small. It answers a preflight itself, and adds its
 * headers to the request that follows. The handler writes x-rb-serial, so its absence on a
 * preflight shows the middleware answered alone.
 */
const corsFamily: Routes = (router, { payloads: p }) => {
  const { origin, method, header, maxAgeSeconds } = p.settings.cors;

  // rb:wiring cors.*
  // The origin as a function that names settings.json's origin only when the request comes from
  // it. Given a string, @koa/cors writes that origin on every answer whichever origin asked, and
  // given a function that returns nothing, it adds no headers and passes the request on.
  const allowed = cors({ origin: (ctx) => (ctx.get("origin") === origin ? origin : ""), allowMethods: [method], allowHeaders: [header], maxAge: maxAgeSeconds });

  // @koa/router runs a route's middleware only for a method the route has, so the preflight has a
  // route of its own, which @koa/cors answers alone.
  // rb:handler cors.preflight
  router.options("/cors/small", allowed);

  // rb:handler cors.request,cors.vary
  router.get("/cors/small", allowed, (ctx) => fresh(ctx, p.small));
};

export default corsFamily;

File: frameworks/python/flask/Implementation/routes/cors.py
from flask import Blueprint, Response, jsonify
# rb:wiring cors.*
from flask_cors import CORS

from payloads import Payloads
from serial import fresh


def blueprint(p: Payloads) -> Blueprint:
    """cors: Flask-CORS on this family's blueprint, which adds its headers after the view, and to the
    OPTIONS answer Flask writes itself for every route. The view writes x-rb-serial, so its absence
    on a preflight shows no view ran."""
    settings = p.settings["cors"]
    routes = Blueprint("cors", __name__)
    # rb:wiring cors.*
    CORS(routes, origins=[settings["origin"]], methods=[settings["method"]], allow_headers=[settings["header"]],
         max_age=settings["maxAgeSeconds"])

    # rb:handler cors.request
    @routes.get("/cors/small")
    def small() -> Response:
        return fresh(jsonify(p.small))

    return routes
